test/helpers.py

```diff
@@ -136,61 +159,270 @@ def _batch_sde_solve(
+    get_dt_step_controller: Callable[
+        [int], tuple[float, diffrax.AbstractStepSizeController]
+    ],
+    saveat: diffrax.SaveAt,
+    bm_tol: float,
 ):
-    dtype = jnp.result_type(*jtu.tree_leaves(y0))
-    # TODO: add a check whether the solver needs levy area
-    levy_area = diffrax.BrownianIncrement
-    keys = jr.split(key, num_samples)  # deliberately reused across all solves
+    levy_area1 = get_minimal_la(solver)
+    levy_area2 = get_minimal_la(ref_solver)
+    # Stricter levy_area requirements inherit from less strict ones
+    levy_area = levy_area1 if issubclass(levy_area1, levy_area2) else levy_area2
+
+    level_coarse, level_fine = levels
 
-    correct_sols = _batch_sde_solve(
+    dt, step_controller = get_dt_step_controller(ref_level)
+    correct_sols, _ = _batch_sde_solve(
         keys,
         get_terms,
-        levy_area,
-        ref_solver,
         w_shape,
         t0,
         t1,
-        dt_precise,
         y0,
         args,
+        ref_solver,
+        levy_area,
+        dt,
+        step_controller,
+        bm_tol,
+        saveat,
     )
-    dts = 2.0 ** jnp.arange(-3, -3 - num_levels, -1, dtype=dtype)
 
-    @jax.jit
-    @jax.vmap
-    def get_single_err(dt):
-        sols = _batch_sde_solve(
+    errs_list, steps_list = [], []
+    for level in range(level_coarse, level_fine + 1):
+        dt, step_controller = get_dt_step_controller(level)
+        sols, steps = _batch_sde_solve(
             keys,
             get_terms,
-            levy_area,
-            solver,
             w_shape,
             t0,
             t1,
-            dt,
             y0,
             args,
+            solver,
+            levy_area,
+            dt,
+            step_controller,
+            bm_tol,
+            saveat,
+        )
+        errs = path_l2_dist(sols, correct_sols)
+        errs_list.append(errs)
+        steps_list.append(jnp.average(steps))
+    errs_arr = jnp.array(errs_list)
+    steps_arr = jnp.array(steps_list)
+    order, _ = jnp.polyfit(jnp.log(1 / steps_arr), jnp.log(errs_arr), 1)
+    return steps_arr, errs_arr, order
+
+
+# TODO: remove this once we have a better way to handle this
+# I understand you'd prefer not to have this in the library and
+# I agree this is somewhat hacky, but I think passing each of
+# these args around separately is a bit of a pain. If this only appeared
+# in the tests, I'd be fine with it, but it's also in the examples
+# (e.g. srk_example.py) and I'd prefer if it looked a bit cleaner there.
+# So how do you recommend we streamline this?
+@dataclasses.dataclass
+class SDE:
+    get_terms: Callable[[AbstractBrownianPath], AbstractTerm]
+    args: PyTree
+    y0: PyTree[Array]
+    t0: float
+    t1: float
+    w_shape: tuple[int, ...]
+
+    def get_dtype(self):
+        return jnp.result_type(*jtu.tree_leaves(self.y0))
+
+    def get_bm(
+        self,
+        bm_key,
+        levy_area: type[diffrax.BrownianIncrement],
+        tol=2**-14,
```

Comment:
I would ask to remove this default argument, though, as I still don't like using default arguments unless you absolutely have to :)

Reply:
Yes, sorry, that was there for backwards-compatibility with various experiment notebooks, and I didn't bother changing it as I wasn't sure whether `SDE` was here to stay. I removed it now.